components/brave_adaptive_captcha/brave_adaptive_captcha_service.cc

```diff
@@ -0,0 +1,189 @@
+#include "services/network/public/cpp/shared_url_loader_factory.h"
+
+namespace brave_adaptive_captcha {
+
+constexpr int kScheduledCaptchaMaxFailedAttempts = 10;
+
+Environment BraveAdaptiveCaptchaService::environment_ = DEVELOPMENT;
+
+const char kScheduledCaptchaId[] = "brave.rewards.scheduled_captcha.id";
+const char kScheduledCaptchaPaymentId[] =
+    "brave.rewards.scheduled_captcha.payment_id";
+const char kScheduledCaptchaSnoozeCount[] =
+    "brave.rewards.scheduled_captcha.snooze_count";
+const char kScheduledCaptchaFailedAttempts[] =
+    "brave.rewards.scheduled_captcha.failed_attempts";
+const char kScheduledCaptchaPaused[] = "brave.rewards.scheduled_captcha.paused";
+
+net::NetworkTrafficAnnotationTag kAnnotationTag =
+    net::DefineNetworkTrafficAnnotation("brave_adaptive_captcha_service", R"(
+        semantics {
+          sender:
+            "Brave Adaptive Captcha service"
+          description:
+            "Fetches CAPTCHA data from Brave."
+          trigger:
+            "The Brave service indicates that it's time to solve a CAPTCHA."
+          data: "Brave CAPTCHA data."
+          destination: WEBSITE
+        }
+        policy {
+          cookies_allowed: NO
+          setting:
+            "This feature cannot be disabled by settings."
+          policy_exception_justification:
+            "Not implemented."
+        })");
+
+BraveAdaptiveCaptchaService::BraveAdaptiveCaptchaService(
+    content::BrowserContext* context,
+    PrefService* prefs,
+    scoped_refptr<network::SharedURLLoaderFactory> url_loader_factory,
+    brave_rewards::RewardsService* rewards_service,
+    std::unique_ptr<BraveAdaptiveCaptchaDelegate> delegate)
+    : prefs_(prefs),
+      rewards_service_(rewards_service),
+      delegate_(std::move(delegate)),
+      api_request_helper_(kAnnotationTag, std::move(url_loader_factory)),
+      captcha_challenge_(
+          std::make_unique<GetAdaptiveCaptchaChallenge>(&api_request_helper_)) {
+  DCHECK(context);
+  DCHECK(prefs);
+  DCHECK(rewards_service);
+
+  rewards_service_->AddObserver(this);
+}
+
+BraveAdaptiveCaptchaService::~BraveAdaptiveCaptchaService() {
+  rewards_service_->RemoveObserver(this);
+}
+
+void BraveAdaptiveCaptchaService::GetScheduledCaptcha(
+    const std::string& payment_id,
+    OnGetAdaptiveCaptchaChallenge callback) {
+  captcha_challenge_->Request(environment_, payment_id, std::move(callback));
+}
+
+bool BraveAdaptiveCaptchaService::GetScheduledCaptchaInfo(
+    std::string* url,
+    bool* max_attempts_exceeded) {
+  DCHECK(url);
+  DCHECK(max_attempts_exceeded);
+
+  const std::string payment_id = prefs_->GetString(kScheduledCaptchaPaymentId);
+  const std::string captcha_id = prefs_->GetString(kScheduledCaptchaId);
+  if (payment_id.empty() || captcha_id.empty()) {
+    return false;
+  }
+
+  const int failed_attempts =
+      prefs_->GetInteger(kScheduledCaptchaFailedAttempts);
+
+  *url = brave_adaptive_captcha::BraveAdaptiveCaptchaService::
```

Comment:
nit: this doesn't need to be fully qualified here

Reply:
Fixed.